### PR TITLE
[Git] Drop most test files & unsupported features from builds to reduce size

### PR DESCRIPTION
**Pull request by andrerom**
Drops most tests and unsupported features _(old broken elasticsearch 1.5 search prototype, this is removed in master and will likely come re-written next year as new package instead)_.

**Main motivation:**
A step to reduce size of install in order to in the end enable use on setups such as AWS Lambda. Same as done for Symfony 4.4 recently here: https://github.com/symfony/symfony/pull/33579

**Size change:**
- 7.5: 5.9mb compressed (download), 41.6mb extracted (disk)
- this: 3.7mb compressed (download), 22.8mb extracted (disk)

_Tested using `git archive --prefix=test/ -o test.zip HEAD`_

⚠️ **Impact:**
Community bundles that relies on Test classes and fixtures in kernel will for testing have to force Composer to install using sources instead of tags (`--prefer-source`), at least for this package. 

**Todo:**
- Test against SolrBundle _(relies on integration test)_
- Get Community to test their packages that rely on kernel test classes
- **TBD:** Identify a minimal safe change for LTS branches and more aggressive stripping for master(?)

### CODE REVIEW

File: .gitattributes [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+/eZ/Publish/Core/REST/Client/Tests export-ignore
+/eZ/Publish/Core/REST/Common/Tests export-ignore
+/eZ/Publish/Core/REST/Server/Tests export-ignore
+
+/eZ/Publish/Core/Search/Legacy/Tests export-ignore
+/eZ/Publish/API/Repository/Tests/_fixtures/Legacy export-ignore
+/eZ/Publish/Core/Search/Tests export-ignore
+
+/eZ/Publish/SPI/Tests export-ignore
```
Comment (alongosz) [inline .gitattributes]: Note to self: I think we really need to separate tests into one common directory :upside_down_face:

File: .gitattributes [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+# Unsupported Elastic Search 1.x prototype
+/eZ/Publish/Core/Search/Elasticsearch export-ignore
+/eZ/Publish/API/Repository/Tests/_fixtures/Elasticsearch export-ignore
```
Comment (alongosz) [inline .gitattributes]: those 2 directories no longer exist in `master`

File: .gitattributes [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+*.php text
+composer.lock merge=ours
+
+# Root
+/.gitattributes export-ignore
+/.travis.yml export-ignore
+/.github export-ignore
+/var export-ignore
+/bisect_phpunit.sh export-ignore
+/bootstrap.php export-ignore
```
Comment (alongosz) [inline .gitattributes]: So if `/config.php-DEVELOPMENT` is commented out, why this one isn't? It's also needed for tests.

**Review by alongosz — CHANGES_REQUESTED** [review]

File: .gitattributes (hunk 2 above)
Comment (andrerom) [reply .gitattributes]: true, let's discuss what is safe to do in 2.5 here somewhere else. This like this I'd like there as it's definitely not supported.

File: .gitattributes (hunk 3 above)
Comment (andrerom) [reply .gitattributes]: This one is needed for kernel tests, but afaik not for solr tests which is the case for `config.php-DEVELOPMENT` or?

That said I opened this originally on 7.5, in this case we can aim to clean this up in master to not need this file anymore at all.

Comment (alongosz) [reply .gitattributes]: > `boostrap.php` one is needed for kernel package tests itself, but not for solr bundle integration tests which only needs `config.php-DEVELOPMENT` afaik.
> 
> That said I opened this originally on 7.5, in this case we can aim to clean this up in master to not need this file anymore at all.

Nah, the point with Solr makes sense :+1: 
I'll try to revise test setup improvements for 8.0 anyway and then we can tackle it.

**Review by alongosz — APPROVED** [review]

File: .gitattributes (hunk 1 above)
Comment (andrerom) [reply .gitattributes]: 👍 
At least per lib and per bundle, see the Symfony PR link.

**Comment by andrerom** [comment]
@pspanja / @emodric Any clear dependencies on test classes / fixtures on your sides?

**Comment by emodric** [comment]
@andrerom I know that at least Tags Bundle uses test classes from kernel. This can probably be remedied by using source as the preferred install.

**Comment by gggeek** [comment]
Agree: any downstream that needs Test stuff from Kernel should be able to set up composer to do a Source install in order to get all the goodies

**Comment by andrerom** [comment]
ok, I'll look into the todos here and wrap this up if any changes are needed. Thanks for the feedback so far!

🗣 If anyone sees this differently, please let us know!

**Comment by andrerom** [comment]
Finally got around to test archive from this on solr bundle, it fails on first test with:
```

1) eZ\Publish\API\Repository\Tests\Values\User\Limitation\ContentTypeLimitationTest::testContentTypeLimitationAllow
Cannot create a repository with predefined user. Check the UserService or RoleService implementation. 

Exception: Symfony\Component\DependencyInjection\Exception\RuntimeException: Please add the class to service "ezpublish.api.storage_engine.legacy.connection" even if it is constructed by a factory since we might need to add method calls based on compile-time checks. in /Users/andrerom/Sites/git/ezplatform-solr-search-engine/vendor/symfony/symfony/src/Symfony/Component/DependencyInjection/Compiler/CheckDefinitionValidityPass.php:48
Stack trace:
#0 /Users/andrerom/Sites/git/ezplatform-solr-search-engine/vendor/symfony/symfony/src/Symfony/Component/DependencyInjection/Compiler/Compiler.php(140): Symfony\Component\DependencyInjection\Compiler\CheckDefinitionValidityPass->process(Object(Symfony\Component\DependencyInjection\ContainerBuilder))
#1 /Users/andrerom/Sites/git/ezplatform-solr-search-engine/vendor/symfony/symfony/src/Symfony/Component/DependencyInjection/ContainerBuilder.php(789): Symfony\Component\DependencyInjection\Compiler\Compiler->compile(Object(Symfony\Component\DependencyInjection\ContainerBuilder))
#2 /Users/andrerom/Sites/git/ezplatform-solr-search-engine/vendor/ezsystems/ezpublish-kernel/eZ/Publish/Core/Base/ServiceContainer.php(190): Symfony\Component\DependencyInjection\ContainerBuilder->compile()
#3 /Users/andrerom/Sites/git/ezplatform-solr-search-engine/vendor/ezsystems/ezpublish-kernel/eZ/Publish/Core/Base/ServiceContainer.php(139): eZ\Publish\Core\Base\ServiceContainer->getContainer()
#4 /Users/andrerom/Sites/git/ezplatform-solr-search-engine/vendor/ezsystems/ezpublish-kernel/eZ/Publish/Core/Base/ServiceContainer.php(83): eZ\Publish\Core\Base\ServiceContainer->initializeContainer()
#5 /Users/andrerom/Sites/git/ezplatform-solr-search-engine/vendor/ezsystems/ezpublish-kernel/eZ/Publish/API/Repository/Tests/SetupFactory/Legacy.php(481): eZ\Publish\Core\Base\ServiceContainer->__construct(Object(Symfony\Component\DependencyInjection\ContainerBuilder), '/Users/andrerom...', '/Users/andrerom...', true, true)
#6 /Users/andrerom/Sites/git/ezplatform-solr-search-engine/vendor/ezsystems/ezpublish-kernel/eZ/Publish/API/Repository/Tests/SetupFactory/Legacy.php(437): eZ\Publish\API\Repository\Tests\SetupFactory\Legacy->getServiceContainer()
#7 /Users/andrerom/Sites/git/ezplatform-solr-search-engine/vendor/ezsystems/ezpublish-kernel/eZ/Publish/API/Repository/Tests/SetupFactory/Legacy.php(341): eZ\Publish\API\Repository\Tests\SetupFactory\Legacy->getDatabaseConnection()
#8 /Users/andrerom/Sites/git/ezplatform-solr-search-engine/vendor/ezsystems/ezpublish-kernel/eZ/Publish/API/Repository/Tests/SetupFactory/Legacy.php(321): eZ\Publish\API\Repository\Tests\SetupFactory\Legacy->createSchema('/Users/andrerom...')
#9 /Users/andrerom/Sites/git/ezplatform-solr-search-engine/vendor/ezsystems/ezpublish-kernel/eZ/Publish/API/Repository/Tests/SetupFactory/Legacy.php(141): eZ\Publish\API\Repository\Tests\SetupFactory\Legacy->initializeSchema()
#10 /Users/andrerom/Sites/git/ezplatform-solr-search-engine/tests/lib/SetupFactory/LegacySetupFactory.php(39): eZ\Publish\API\Repository\Tests\SetupFactory\Legacy->getRepository(false)
#11 /Users/andrerom/Sites/git/ezplatform-solr-search-engine/vendor/ezsystems/ezpublish-kernel/eZ/Publish/API/Repository/Tests/BaseTest.php(55): EzSystems\EzPlatformSolrSearchEngine\Tests\SetupFactory\LegacySetupFactory->getRepository(false)
#12 /Users/andrerom/Sites/git/ezplatform-solr-search-engine/vendor/phpunit/phpunit/src/Framework/TestCase.php(838): eZ\Publish\API\Repository\Tests\BaseTest->setUp()
#13 /Users/andrerom/Sites/git/ezplatform-solr-search-engine/vendor/phpunit/phpunit/src/Framework/TestResult.php(693): PHPUnit\Framework\TestCase->runBare()
#14 /Users/andrerom/Sites/git/ezplatform-solr-search-engine/vendor/phpunit/phpunit/src/Framework/TestCase.php(796): PHPUnit\Framework\TestResult->run(Object(eZ\Publish\API\Repository\Tests\Values\User\Limitation\ContentTypeLimitationTest))
#15 /Users/andrerom/Sites/git/ezplatform-solr-search-engine/vendor/phpunit/phpunit/src/Framework/TestSuite.php(746): PHPUnit\Framework\TestCase->run(Object(PHPUnit\Framework\TestResult))
#16 /Users/andrerom/Sites/git/ezplatform-solr-search-engine/vendor/phpunit/phpunit/src/Framework/TestSuite.php(746): PHPUnit\Framework\TestSuite->run(Object(PHPUnit\Framework\TestResult))
#17 /Users/andrerom/Sites/git/ezplatform-solr-search-engine/vendor/phpunit/phpunit/src/Framework/TestSuite.php(746): PHPUnit\Framework\TestSuite->run(Object(PHPUnit\Framework\TestResult))
#18 /Users/andrerom/Sites/git/ezplatform-solr-search-engine/vendor/phpunit/phpunit/src/TextUI/TestRunner.php(652): PHPUnit\Framework\TestSuite->run(Object(PHPUnit\Framework\TestResult))
#19 /Users/andrerom/Sites/git/ezplatform-solr-search-engine/vendor/phpunit/phpunit/src/TextUI/Command.php(206): PHPUnit\TextUI\TestRunner->doRun(Object(PHPUnit\Framework\TestSuite), Array, true)
#20 /Users/andrerom/Sites/git/ezplatform-solr-search-engine/vendor/phpunit/phpunit/src/TextUI/Command.php(162): PHPUnit\TextUI\Command->run(Array, true)
#21 /Users/andrerom/Sites/git/ezplatform-solr-search-engine/vendor/phpunit/phpunit/phpunit(61): PHPUnit\TextUI\Command::main()
#22 {main}

/Users/andrerom/Sites/git/ezplatform-solr-search-engine/vendor/ezsystems/ezpublish-kernel/eZ/Publish/API/Repository/Tests/BaseTest.php:74

FAILURES!
Tests: 1, Assertions: 1, Failures: 1.
```


@alongosz and @adamwojs Should we keep this for master only maybe? Or do we assume stable packages like solr fetches kernel from dev/source always?

**Comment by alongosz** [comment]
> Finally got around to test archive from this on solr bundle, it fails on first test with:
> ```
> 
> 1) eZ\Publish\API\Repository\Tests\Values\User\Limitation\ContentTypeLimitationTest::testContentTypeLimitationAllow
> Cannot create a repository with predefined user. Check the UserService or RoleService implementation. 
> 
> Exception: Symfony\Component\DependencyInjection\Exception\RuntimeException: Please add the class to service "ezpublish.api.storage_engine.legacy.connection" even if it is constructed by a factory since we might need to add method calls based on compile-time checks.
> (...)
> ```
>  Should we keep this for master only maybe? Or do we assume stable packages like solr fetches kernel from dev/source always?

@andrerom the fact that test setup from Solr relies on test setup from Kernel is a mistake. Usually those namespaces are `autoload-dev` (obviously not the case for Kernel), so they shouldn't even be accessible. So, no, I would not make such assumption. The direction here should be rather fixing test setup for Kernel and Solr. 
The only thing that should be exposed from Kernel is SetupFactory (as the test API), other portions of test matrix should not depend on Kernel. They could depend on some internal dev package though if we see that there's no way to keep it DRY otherwise.

Though, I'm not following how the mentioned failure is related to reduced build. Are some classes missing which results in `ezpublish.api.storage_engine.legacy.connection` not being built?

**Comment by andrerom** [comment]
Actually I'm not entirely sure. Might actually be unrelated.
I have limited time to look at this so if anyone feels like don't hesitate to take over, otherwise I'll try to dig into that at a later point.

**Pull request closed without merging**